src/raftstore/store/worker/apply.rs

```diff
@@ -550,51 +592,67 @@ impl ApplyDelegate {
         self.id
     }
 
-    fn handle_raft_committed_entries(
-        &mut self,
-        apply_ctx: &mut ApplyContext,
+    pub fn handle_raft_committed_entries(
+        self,
+        mut apply_ctx: ApplyContext,
         committed_entries: Vec<Entry>,
-    ) {
+    ) -> Box<Future<Item = (Self, ApplyContext), Error = (Self, ApplyContext, Error)> + Send> {
         if committed_entries.is_empty() {
-            return;
+            return box future::ok((self, apply_ctx));
         }
-        apply_ctx.prepare_for(self);
+        apply_ctx.prepare_for(&self);
         apply_ctx.committed_count += committed_entries.len();
         // If we send multiple ConfChange commands, only first one will be proposed correctly,
         // others will be saved as a normal entry with no data, so we must re-propose these
         // commands again.
         apply_ctx.committed_count += committed_entries.len();
-        let mut results = vec![];
-        for entry in committed_entries {
-            if self.pending_remove {
-                // This peer is about to be destroyed, skip everything.
-                break;
-            }
 
-            let expect_index = self.apply_state.get_applied_index() + 1;
-            if expect_index != entry.get_index() {
-                panic!(
-                    "{} expect index {}, but got {}",
-                    self.tag,
-                    expect_index,
-                    entry.get_index()
-                );
-            }
+        let f = futures::stream::iter_ok(committed_entries).fold(
```

Comment:
`fold` needs to return an `Error` to end the loop ahead of time. So I think `loop_fn` is better.